File: Java/net.alcetech.UniversalEditor.Plugins.Core/src/net/alcetech/UniversalEditor/DataFormats/Markup/XML/XMLDataFormat.java
package net.alcetech.UniversalEditor.DataFormats.Markup.XML;

import java.io.IOException;

import net.alcetech.ApplicationFramework.Collections.Generic.KeyValuePair;
import net.alcetech.ApplicationFramework.Exceptions.*;
import net.alcetech.UniversalEditor.Core.*;
import net.alcetech.UniversalEditor.Core.IO.Reader;
import net.alcetech.UniversalEditor.Core.IO.SeekOrigin;
import net.alcetech.UniversalEditor.ObjectModels.Markup.*;

public class XMLDataFormat extends DataFormat
{
	private static DataFormatReference _dfr = null;
	public DataFormatReference getDataFormatReference() {
		if (_dfr == null)
		{
			_dfr = super.getDataFormatReference();
			_dfr.getSupportedObjectModelCollection().add(ObjectModelReference.fromClass(MarkupObjectModel.class));
		}
		return _dfr;
	}
	
	private XMLDataFormatSettings _settings = new XMLDataFormatSettings();
	public XMLDataFormatSettings getSettings() { return _settings; }
	
	public XMLDataFormat() {
		// this.getSettings().addAutoCloseTagName("hr");
		// this.getSettings().addAutoCloseTagName("br");
		// this.getSettings().addAutoCloseTagName("img");
		this.getSettings().addEntity("amp", "&");
		this.getSettings().addEntity("quot", "\"");
		this.getSettings().addEntity("copy", "©");
		this.getSettings().addEntity("apos", "'");
	}
	
	@Override
	protected void loadInternal(ObjectModel objectModel) throws IOException {
		Reader tr = super.getAccessor().getReader();
		
		MarkupObjectModel mom = new MarkupObjectModel();
		boolean insideTagDecl = false;
		int insideString = 0;
		boolean closingCurrentElement = false;
		String currentString = "";
		MarkupElement currentElement = null;
		boolean insidePreprocessor = false;
		boolean insideAttributeArea = false;
		boolean insideSpecialDeclaration = false;
		boolean insideTagValue = false;
		char prevChar = '\0';
		String nextAttributeName = null;
		boolean loaded = false;

		char c = tr.readChar();
		int times = 0, maxtimes = 5;
		while (c != '<')
		{
			// clear out junk
			c = tr.readChar();
			times++;
			if (times == maxtimes) break;
		}
		tr.getAccessor().seek(-1, SeekOrigin.CURRENT);

		boolean seenBeginChar = false;
		
		while (!tr.getEndOfStream())
		{
			c = tr.readChar();
			if (!loaded && (c != '<'))
			{
				return;
			}
			else
			{
				loaded = true;
			}

			if (c == this.getSettings().getTagEndChar() && !seenBeginChar)
			{
				currentString += c;
				continue;
			}
			else if (c == this.getSettings().getTagEndChar() && seenBeginChar)
			{
				seenBeginChar = false;
			}

			if (c == this.getSettings().getTagBeginChar())
			{
				if (insideString == 0 && !insidePreprocessor)
				{
					insideTagDecl = true;
					insideTagValue = false;
					if (!("".equals(currentString)) && currentElement != null)
					{
						currentElement.setValue(currentElement.getValue() + currentString);
						currentElement.setValue(this.replaceEntitiesInput(currentElement.getValue().trim()));
						currentString = "";
					}
				}
				else
				{
					currentString += c;
				}
				seenBeginChar = true;
			}
			else
			{
				if (c == this.getSettings().getTagSpecialDeclarationStartChar())
				{
					if (insideString == 0 && !insidePreprocessor && !insideTagValue && !insideAttributeArea && prevChar == this.getSettings().getTagBeginChar())
					{
						insideSpecialDeclaration = true;
						continue;
					}
					currentString += c;
				}
				else
				{
					if (c == this.getSettings().getTagEndChar())
					{
						if (insideString == 0 && ((insidePreprocessor && prevChar == this.getSettings().getPreprocessorChar()) || prevChar != this.getSettings().getPreprocessorChar()))
						{
							insideTagDecl = false;
							if (currentString.endsWith(this.getSettings().getTagSpecialDeclarationCommentStart()))
							{
								currentString = currentString.substring(0, currentString.length() - this.getSettings().getTagSpecialDeclarationCommentStart().length());
								currentElement.setValue(this.replaceEntitiesInput(currentString));
								currentString = "";
								// currentElement = currentElement.getParentElement();
								continue;
							}
							if (insidePreprocessor && prevChar == this.getSettings().getPreprocessorChar())
							{
								MarkupElement prevElement = currentElement;
								String name = "";
								String content = currentString;
								if (currentString.contains(" "))
								{
									name = currentString.substring(0, currentString.indexOf(' '));
									content = currentString.substring(name.length() + 1);
								}
								currentElement = new MarkupPreprocessorElement(name, content);
								if (prevElement != null && prevElement.getClass().isAssignableFrom(MarkupTagElement.class))
								{
									((MarkupTagElement)prevElement).addElement(currentElement);
								}
								else
								{
									mom.addElement(currentElement);
								}
								currentElement = prevElement;
								insidePreprocessor = false;
							}
							else
							{
								if (closingCurrentElement)
								{
									if (currentElement != null)
									{
										if (currentElement.getFullName().equals(currentString) || prevChar == this.getSettings().getTagCloseChar())
										{
											currentElement = currentElement.getParentElement();
										}
										else
										{
											if (this.getSettings().containsAutoCloseTagName(currentElement.getName()))
											{
												currentElement = currentElement.getParentElement();
												if (currentElement != null)
												{
													currentElement = currentElement.getParentElement();
												}
											}
											else
											{
												System.err.println("uds: XMLDataFormat: in " + this.getAccessor().getFileName());
												System.err.println("uds: XMLDataFormat: attempted to close element '" + currentElement.getName() + "' with '" + currentString + "'; invalid XML?");
											}
										}
									}
									else
									{
										// Console.WriteLine("FIXME: UniversalDataStorage/FileFormats/Markup/XMLFileFormat.cs:144     currentElement is null");
										System.err.println("FIXME: UniversalEditor/DataFormats/Markup/XML/XMLDataFormat.java:184     currentElement is null");
									}
									closingCurrentElement = false;
								}
								else
								{
									if (insideSpecialDeclaration)
									{
										insideSpecialDeclaration = false;
									}
									else
									{
										insideTagValue = true;
										if (insideAttributeArea)
										{
											((MarkupTagElement)currentElement).addAttribute(nextAttributeName, this.replaceEntitiesInput(currentString));
											nextAttributeName = null;
											insideAttributeArea = false;
										}
										else
										{
											MarkupElement prevElement = currentElement;
											currentElement = new MarkupTagElement(currentString.trim());
											if (prevElement != null && prevElement.getClass().isAssignableFrom(MarkupTagElement.class))
											{
												((MarkupTagElement)prevElement).addElement(currentElement);
											}
											else
											{
												mom.addElement(currentElement);
											}
											if (this.getSettings().containsAutoCloseTagName(currentElement.getName()))
											{
												currentElement = currentElement.getParentElement();
											}
										}
									}
								}
							}
							currentString = "";
							insideAttributeArea = false;
						}
						else
						{
							currentString += c;
						}
					}
					else
					{
						if (c == this.getSettings().getTagCloseChar())
						{
							if (insideString == 0 && !insidePreprocessor && insideTagDecl)
							{
								if (insideAttributeArea && currentElement.getClass().isAssignableFrom(MarkupTagElement.class) && nextAttributeName != null)
								{
									if (nextAttributeName != null)
									{
										((MarkupTagElement)currentElement).addAttribute(nextAttributeName, this.replaceEntitiesInput(currentString));
										nextAttributeName = null;
										currentString = "";
									}
								}
								else
								{
									if (!insideAttributeArea)
									{
										if (currentElement != null)
										{
											if (currentElement.getClass().isAssignableFrom(MarkupTagElement.class) && !"".equals(currentString))
											{
												MarkupTagElement e = new MarkupTagElement(currentString);
												((MarkupTagElement)currentElement).addElement(e);
												currentElement = e;
											}
										}
										else
										{
											// Console.WriteLine("FIXME: UniversalDataStorage/FileFormats/Markup/XMLFileFormat.cs:199     currentElement is null");
											System.out.println("FIXME: UniversalEditor/DataFormats/Markup/XML/XMLDataFormat.java:265     currentElement is null");
										}
									}
								}
								currentString = "";
								closingCurrentElement = true;
							}
							else
							{
								currentString += c;
							}
						}
						else
						{
							if (c == this.getSettings().getAttributeNameValueSeparatorChar())
							{
								if (insideString == 0 && !insidePreprocessor && !insideTagValue)
								{
									if (insideAttributeArea)
									{
										nextAttributeName = currentString;
										currentString = "";
										char ccq = tr.peekChar();
										if (ccq != '"' && ccq != '\'')
										{
											char cc = '\0';
											String nnn = "";
											while (cc != this.getSettings().getTagEndChar())
											{
												cc = tr.peekChar();
												if (cc == this.getSettings().getTagEndChar())
												{
													break;
												}
												nnn += tr.readChar();
												if (tr.getEndOfStream())
												{
													break;
												}
											}
											String value = nnn;
											if (nnn.contains(String.valueOf(this.getSettings().getAttributeNameValueSeparatorChar())))
											{
												int l = nnn.indexOf(' ');
												value = nnn.substring(0, l);
												String nnnn = nnn.substring(value.length()).trim();
												String nnnnAttributeName = "";
												String nnnnAttributeValue = "";
												boolean nnnnLookForValue = false;
												int nnnnIsInsideString = 0;
												for (int i = 0; i < nnnn.length(); i++)
												{
													char r = nnnn.charAt(i);
													if (r == this.getSettings().getAttributeNameValueSeparatorChar())
													{
														if (nnnnIsInsideString == 0)
														{
															nnnnLookForValue = true;
														}
													}
													else
													{
														if (r == '\'' || r == '"')
														{
															if ((r == '\'' && nnnnIsInsideString == 2) || (r == '"' && nnnnIsInsideString == 1))
															{
																nnnnIsInsideString = 0;
															}
															else
															{
																if (r == '"')
																{
																	nnnnIsInsideString = 1;
																}
																else
																{
																	if (r == '\'')
																	{
																		nnnnIsInsideString = 2;
																	}
																}
															}
														}
														else
														{
															if (r == ' ' && nnnnIsInsideString == 0)
															{
																if (currentElement != null)
																{
																	if (currentElement.getClass().isAssignableFrom(MarkupTagElement.class))
																	{
																		((MarkupTagElement)currentElement).addAttribute(nnnnAttributeName, this.replaceEntitiesInput(nnnnAttributeValue));
																		nnnnAttributeName = "";
																		nnnnAttributeValue = "";
																		nnnnLookForValue = false;
																	}
																}
															}
															else
															{
																if (nnnnLookForValue)
																{
																	nnnnAttributeValue += r;
																}
																else
																{
																	nnnnAttributeName += r;
																}
															}
														}
													}
												}
												if (!("".equals(nnnnAttributeName)))
												{
													((MarkupTagElement)currentElement).addAttribute(nnnnAttributeName, this.replaceEntitiesInput(nnnnAttributeValue));
													nnnnAttributeName = "";
													nnnnAttributeValue = "";
												}
											}
											if (currentElement != null)
											{
												if (currentElement.getClass().isAssignableFrom(MarkupTagElement.class))
												{
													((MarkupTagElement)currentElement).addAttribute(nextAttributeName, this.replaceEntitiesInput(value));
													nextAttributeName = null;
												}
											}
											currentString = "";
											continue;
										}
									}
								}
								else
								{
									currentString += c;
								}
							}
							else
							{
								if (c == '"' || c == '\'')
								{
									if (!insidePreprocessor)
									{
										if (!insideTagDecl)
										{
											currentString += c;
											continue;
										}
										int cur = 0;
										if (c == '"')
										{
											cur = 1;
										}
										if (c == '\'')
										{
											cur = 2;
										}
										if (insideString == 0)
										{
											insideString = cur;
										}
										else
										{
											if (insideString == cur)
											{
												insideString = 0;
											}
											else
											{
												currentString += c;
											}
										}
									}
									else
									{
										currentString += c;
									}
								}
								else
								{
									if (Character.isWhitespace(c))
									{
										if (insideString == 0 && !insidePreprocessor)
										{
											if (Character.isWhitespace(prevChar))
											{
												continue;
											}
											if (insideTagDecl)
											{
												if (insideSpecialDeclaration)
												{
													if (currentString.trim() == "DOCTYPE")
													{
														String doctype = tr.readStringUntil(this.getSettings().getTagEndChar(), "\"", "\"");
														doctype = doctype.substring(0, doctype.length() - 1);
														/*
														String[] doctypeParts = doctype.split(new String[]
														{
															" "
														}, "\"", "\"");
														*/
														insideTagDecl = false;
														insideSpecialDeclaration = false;
													}
												}
												else
												{
													if (!insideAttributeArea)
													{
														MarkupElement prevElement = currentElement;
														currentElement = new MarkupTagElement(currentString.trim());
														if (prevElement == null)
														{
															mom.addElement(currentElement);
														}
														else
														{
															if (prevElement.getClass().isAssignableFrom(MarkupTagElement.class))
															{
																((MarkupTagElement)prevElement).addElement(currentElement);
															}
														}
														insideAttributeArea = true;
													}
													else
													{
														if (nextAttributeName != null)
														{
															if (currentElement != null && (currentElement.getClass().isAssignableFrom(MarkupTagElement.class)))
															{
																((MarkupTagElement)currentElement).addAttribute(nextAttributeName, this.replaceEntitiesInput(currentString));
																nextAttributeName = null;
															}
														}
													}
												}
												currentString = "";
											}
											else
											{
												if (currentString.length() == 0 || !Character.isWhitespace(currentString.charAt(currentString.length() - 1)))
												{
													currentString += ' ';
												}
											}
										}
										else
										{
											if (insidePreprocessor || currentString.length() == 0 || !Character.isWhitespace(currentString.charAt(currentString.length() - 1)))
											{
												currentString += c;
											}
										}
									}
									else
									{
										if (c == this.getSettings().getPreprocessorChar() && insideString == 0 && insideTagDecl)
										{
											if (prevChar == this.getSettings().getTagBeginChar())
											{
												insidePreprocessor = true;
											}
										}
										else if (insideSpecialDeclaration && c == this.getSettings().getCDataBeginChar())
										{
											String specialSectionName = tr.readStringUntil(this.getSettings().getCDataBeginChar());
											specialSectionName = specialSectionName.substring(0, specialSectionName.length());
											tr.getAccessor().seek(1, SeekOrigin.CURRENT);

											String specialSectionContent = tr.readStringUntil(this.getSettings().getCDataEndChar());
											if (specialSectionContent.length() > 0)
											{
												specialSectionContent = specialSectionContent.substring(0, specialSectionContent.length()); 
											}

											MarkupStringElement tag = new MarkupStringElement(specialSectionName, specialSectionContent);
											if (currentElement != null && (currentElement.getClass().isAssignableFrom(MarkupTagElement.class)))
											{
												((MarkupTagElement)currentElement).addElement(tag);
											}
											else
											{
												mom.addElement(tag);
											}

											char padNext = tr.readChar();
											padNext = tr.readChar();

											// insideSpecialDeclaration = false;
											continue;
										}
										else
										{
											currentString += c;
											if (insideSpecialDeclaration)
											{
												if (currentString.trim().equals(this.getSettings().getTagSpecialDeclarationCommentStart()))
												{
													currentString = "";
													String commentContent = tr.readStringUntil(this.getSettings().getTagSpecialDeclarationCommentStart() + this.getSettings().getTagEndChar());
													MarkupCommentElement comment = new MarkupCommentElement(commentContent);
													if (currentElement == null)
													{
														mom.addElement(comment);
													}
													else
													{
														if (currentElement.getClass().isAssignableFrom(MarkupTagElement.class))
														{
															((MarkupTagElement)currentElement).addElement(comment);
														}
													}
													insideSpecialDeclaration = false;
													insideTagDecl = false;
												}
											}
											else
											{
												if (insideString != 0 || !Character.isWhitespace(c))
												{
												}
											}
										}
									}
								}
							}
						}
					}
				}
			}
			prevChar = c;
		}

		if (objectModel.getClass().isAssignableFrom(MarkupObjectModel.class))
		{
			mom.copyTo(objectModel);
		}
	}
	@Override
	protected void saveInternal(ObjectModel objectModel) throws IOException {
		// TODO Auto-generated method stub
		
	}
	
	private String replaceEntitiesInput(String value) {
		if (value != "")
		{
			KeyValuePair<String, String>[] ents = this.getSettings().getEntities();
			for (int i = 0; i < ents.length; i++)
			{
				KeyValuePair<String, String> kvp = ents[i];
				value = value.replaceAll(String.valueOf(this.getSettings().getEntityBeginChar()) + kvp.getKey() + String.valueOf(this.getSettings().getEntityEndChar()), kvp.getValue());
			}
		}
		return value;
	}
	private String replaceEntitiesOutput(String value) {
		if (value != "")
		{
			KeyValuePair<String, String>[] ents = this.getSettings().getEntities();
			for (int i = 0; i < ents.length; i++)
			{
				KeyValuePair<String, String> kvp = ents[i];
				value = value.replaceAll(kvp.getValue(), String.valueOf(this.getSettings().getEntityBeginChar()) + kvp.getKey() + String.valueOf(this.getSettings().getEntityEndChar()));
			}
		}
		return value;
	}
}

File: CPlusPlus/net.alcetech.UniversalEditor.Essential/DataFormats/Markup/XML/XMLDataFormat.cpp
/*
 * XMLDataFormat.cpp
 *
 *  Created on: Apr 1, 2016
 *      Author: beckermj
 */

#include "XMLDataFormat.h"

using UniversalEditor::IO::Reader;

using ApplicationFramework::Console;
using ApplicationFramework::Text::StringBuilder;

namespace UniversalEditor {
namespace DataFormats {
namespace Markup {
namespace XML {

XMLDataFormat::XMLDataFormat() {
	this->_settings = new XMLDataFormatSettings();

	// this->getSettings()->getAutoCloseTagNamesList()->add("hr");
	// this->getSettings()->getAutoCloseTagNamesList()->add("br");
	// this->getSettings()->getAutoCloseTagNamesList()->add("img");
	this->getSettings()->getEntitiesDictionary()->add(new String("amp"), new String("&"));
	this->getSettings()->getEntitiesDictionary()->add(new String("quot"), new String("\""));
	this->getSettings()->getEntitiesDictionary()->add(new String("copy"), new String("©"));
	this->getSettings()->getEntitiesDictionary()->add(new String("apos"), new String("'"));
}

XMLDataFormat::~XMLDataFormat() {
	// TODO Auto-generated destructor stub
}

XMLDataFormatSettings* XMLDataFormat::getSettings() {
	return this->_settings;
}

const char* replaceEntitiesInput(const char* value) {
	return value;
}

void XMLDataFormat::loadInternal(ObjectModel* objectModel)
{
	// TODO: figure out why comment <!-- --> doesn't work...

	MarkupObjectModel* mom = (MarkupObjectModel*)objectModel;
	if (mom == NULL)
	{
		// TODO: throw a new ObjectModelNotSupported exception here
		printf("ue: XMLDataFormat: loadInternal() - object model not supported\n");
		return;
	}

	Accessor* acc = this->getAccessor();
	Reader* reader = new Reader(acc);
	XMLDataFormatContext* ctx = new XMLDataFormatContext();

	ctx->setNextChar(reader->readChar());
	int times = 0, maxtimes = 10;
	while (ctx->getNextChar() != '<')
	{
		// clear out junk
		ctx->setNextChar(reader->readChar());
		times++;
		if (times == maxtimes) break;
	}
	acc->seek(-1, SeekOrigin::Current);

	while (!acc->endOfStream())
	{
		ctx->setNextChar(reader->readChar());

		if (ctx->isInsideString() || ctx->isInsidePreprocessor())
		{
			if (ctx->getNextChar() == '"' && ctx->isInsideString())
			{
				ctx->setInsideString(false);
				ctx->continueLoop();
				continue;
			}

			if (ctx->getNextChar() == this->getSettings()->getTagEndChar() && ctx->getPrevChar() == this->getSettings()->getPreprocessorChar() && ctx->isInsidePreprocessor())
			{
				ctx->setInsidePreprocessor(false);

				ctx->getNextString()->set(ctx->getNextString()->substring(0, ctx->getNextString()->getLength() - 1));

				MarkupPreprocessorElement* preproc = new MarkupPreprocessorElement();
				preproc->setName(new String("preproc"));
				preproc->setContent(ctx->getNextString()->toString());

				MarkupTagElement* tag = ctx->getCurrentTag();
				if (tag != NULL)
				{
					printf("adding child MarkupPreprocessorElement to tag\n");
					tag->addChildElement(preproc);
				}
				else
				{
					printf("adding child MarkupPreprocessorElement to mom\n");
					mom->addChildElement(preproc);
				}

				ctx->getNextString()->clear();
				ctx->continueLoop();
				continue;
			}

			ctx->getNextString()->append(ctx->getNextChar());
			ctx->continueLoop();
			continue;
		}
		else
		{
			if (ctx->isWhitespaceNotSpace(ctx->getNextChar())) continue;
		}

		if (ctx->getNextChar() == this->getSettings()->getTagBeginChar())
		{
			ctx->setInsideTag(true);
			ctx->setInsideTagValue(false);
			ctx->continueLoop();
			continue;
		}
		else if (ctx->getNextChar() == this->getSettings()->getTagCloseChar())
		{
			if (ctx->getPrevChar() == this->getSettings()->getTagBeginChar())
			{
				ctx->setInsideTagValue(false);

				StringBuilder* closingTagName = new StringBuilder();
				while (ctx->getNextChar() != this->getSettings()->getTagEndChar())
				{
					ctx->setNextChar(reader->readChar());
					if (ctx->getNextChar() != this->getSettings()->getTagEndChar())
					{
						closingTagName->append(ctx->getNextChar());
					}
				}

				if (!ctx->getNextString()->isEmpty())
				{
					if (ctx->getCurrentTag() != NULL)
					{
						printf("setting tag '%s' value '%s'\n", ctx->getCurrentTag()->getFullName()->toCharArray(), ctx->getNextString()->toString()->toCharArray());
						ctx->getCurrentTag()->setValue(ctx->getNextString()->toString());
						ctx->getNextString()->clear();
					}
				}

				if (ctx->getCurrentElement() != NULL)
				{
					if (closingTagName->equals(ctx->getCurrentElement()->getFullName()))
					{
						ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
					}
					else
					{
						printf("cannot close block tag '%s' with '%s'\n", ctx->getCurrentElement()->getFullName()->toCharArray(), closingTagName->toString()->toCharArray());
					}
				}
				else
				{
					printf("cannot close NULL block tag with '%s'\n", closingTagName->toString()->toCharArray());
				}
			}
			else
			{
				if (ctx->getCurrentElement() != NULL)
				{
					ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
				}
			}
			continue;
		}
		else if (ctx->getNextChar() == this->getSettings()->getTagSpecialDeclarationStartChar())
		{
			ctx->setInsideSpecialDeclaration(true);
			continue;
		}
		else if (ctx->getNextChar() == this->getSettings()->getPreprocessorChar())
		{
			ctx->setInsidePreprocessor(true);
		}
		else if (ctx->getNextChar() == this->getSettings()->getAttributeNameValueSeparatorChar())
		{
			if (ctx->isInsideAttributeArea())
			{
				ctx->setNextAttribute(new MarkupAttribute());

				// i feel like we set this multiple times... once here and once in comparing with getTagEndChar
				// why the hell...??? it doesn't work without this one here though
				ctx->getNextAttributeName()->set(ctx->getNextString()->toString());

				ctx->getNextAttribute()->setName(ctx->getNextString()->toString());

				MarkupTagElement* tag = ctx->getCurrentTag();
				if (tag != NULL)
				{
					tag->addAttribute(ctx->getNextAttribute());
				}
				else
				{
					printf("tag is NULL\n");
				}
				ctx->getNextString()->clear();
				continue;
			}
		}
		else if (ctx->getNextChar() == this->getSettings()->getTagEndChar())
		{
			if (ctx->getNextString()->endsWith(this->getSettings()->getTagSpecialDeclarationCommentStart()))
			{

				/*
				printf("next string '%s' is comment\n", ctx->getNextString()->toString()->toCharArray());
				ctx->getNextString()->set(ctx->getNextString()->substring(0, ctx->getNextString()->getLength() - strlen(this->getSettings()->getTagSpecialDeclarationCommentStart())));

				MarkupCommentElement* comment = dynamic_cast<MarkupCommentElement*>(ctx->getCurrentElement());
				if (comment != NULL)
				{
					comment->setContent(ctx->getNextString()->toString());
					printf("adding comment '%s'\n", comment->getContent()->toCharArray());
				}
				else {
					printf("not found comment for '%s'\n", ctx->getNextString()->toString()->toCharArray());
				}
				ctx->getNextString()->clear();

				ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
				*/
			}
			else
			{
				if (ctx->isClosingCurrentElement())
				{
					if (ctx->getCurrentElement() != NULL)
					{
						if (ctx->getPrevChar() == this->getSettings()->getTagCloseChar())
						{
							ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
						}
						else
						{
							if (this->getSettings()->getAutoCloseTagNamesList()->contains(ctx->getCurrentElement()->getName()))
							{
								ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
								if (ctx->getCurrentElement() != NULL)
								{
									ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
								}
							}
							else
							{
								Console::writeLine("uds: XMLDataFormat: attempted to close element '%s'; invalid XML?", ctx->getCurrentElement()->getName()->toCharArray());
							}
						}
					}
					else
					{
						Console::writeLine("FIXME: UniversalDataStorage/FileFormats/Markup/XMLFileFormat.cs:144     currentElement is null");
					}
					ctx->setClosingCurrentElement(false);
				}
				else
				{
					if (ctx->isInsideSpecialDeclaration())
					{
						ctx->setInsideSpecialDeclaration(false);
					}
					else
					{
						ctx->setInsideTagValue(true);
						if (ctx->isInsideAttributeArea())
						{
							// TODO: fix this, see ".externalToolBuilders/Copy all files to share.launch"
							MarkupTagElement* tag = ctx->getCurrentTag();
							if (tag != NULL)
							{
								printf("adding to tag '%s' attribute %s = '%s'\n", tag->getFullName()->toCharArray(), ctx->getNextAttributeName()->toString()->toCharArray(), ctx->getNextString()->toString()->toCharArray());
								tag->addAttribute(ctx->getNextAttributeName()->toString(), ctx->getNextString()->toString());
							}
							ctx->getNextAttributeName()->clear();

							// OOPS: I thought this would fix #2694 but it doesn't
							// ctx->getNextString()->clear();
							ctx->setInsideAttributeArea(false);
						}
						else
						{
							MarkupElement* prevElement = ctx->getCurrentElement();

							MarkupTagElement* tag = new MarkupTagElement();
							ctx->setCurrentElement(tag);
							ctx->getCurrentElement()->setFullName(ctx->getNextString()->toString());
							ctx->getNextString()->clear();

							MarkupTagElement* prevTag = dynamic_cast<MarkupTagElement*>(prevElement);
							if (prevTag != NULL)
							{
								prevTag->addChildElement(ctx->getCurrentElement());
							}
							else
							{
								mom->addChildElement(ctx->getCurrentElement());
							}
							if (this->getSettings()->getAutoCloseTagNamesList()->contains(ctx->getCurrentElement()->getName()))
							{
								ctx->setCurrentElement(ctx->getCurrentElement()->getParentElement());
							}
							continue;
						}
					}
				}
			}

			if (!ctx->getNextString()->isEmpty())
			{
				if (ctx->getNextAttribute() != NULL)
				{
					ctx->getNextAttribute()->setValue(ctx->getNextString()->toString());

					// hack: (2694) we really need to figure out why NextAttribute is getting its value set twice
					// but in order to prevent it we just set it to NULL after the value has been set
					ctx->setNextAttribute(NULL);
				}
			}

			ctx->getNextString()->clear();
			ctx->setInsideAttributeArea(false);
			continue;
		}

		switch (ctx->getNextChar())
		{
			case '"':
			{
				if (!ctx->isInsideString()) ctx->setInsideString(true);
				break;
			}
			case ' ':
			{
				if (ctx->isInsideAttributeArea())
				{
					if (ctx->getNextAttribute() != NULL)
					{
						ctx->getNextAttribute()->setValue(ctx->getNextString()->toString());

						ctx->getNextString()->clear();
						ctx->setNextAttribute(NULL);
					}
				}
				else if (ctx->isInsideTagValue())
				{
					ctx->getNextString()->append(' ');
					continue;
				}
				else
				{
					if (ctx->getNextString()->startsWith(this->getSettings()->getTagSpecialDeclarationCommentStart()))
					{
						StringBuilder* sbCommentEnd = new StringBuilder();
						sbCommentEnd->append(this->getSettings()->getTagSpecialDeclarationCommentStart());
						sbCommentEnd->append(this->getSettings()->getTagEndChar());
						String* commentEndStr = sbCommentEnd->toString();
						ctx->getNextString()->append(' ');
						while (!ctx->getNextString()->endsWith(commentEndStr))
						{
							ctx->setNextChar(reader->readChar());
							ctx->getNextString()->append(ctx->getNextChar());
						}

						printf("%s\n", ctx->getNextString()->toString()->toCharArray());

						ctx->getNextString()->trim();

						/*
						int start = strlen(this->getSettings()->getTagSpecialDeclarationCommentStart());
						int length = ctx->getNextString()->getLength() - start;
						ctx->getNextString()->set(ctx->getNextString()->substring(start, length));

						MarkupCommentElement* comment = new MarkupCommentElement();
						comment->setContent(ctx->getNextString()->toString());
						if (ctx->getCurrentElement() != NULL)
						{
							MarkupTagElement* tag = ctx->getCurrentTag();
							if (tag != NULL)
							{
								printf("added comment '%s' to element '%s'\n", comment->getContent()->toCharArray(), tag->getFullName()->toCharArray());
								tag->addChildElement(comment);
							}
							else
							{
								printf("ue: XMLDataFormat - attempted to add a MarkupCommentElement to a non-Tag element\n");
							}
						}
						else
						{
							printf("added comment '%s' to mom\n", comment->getContent()->toCharArray());
							mom->addChildElement(comment);
						}

						*/
						ctx->getNextString()->clear();
						continue;
					}

					MarkupTagElement* tag = new MarkupTagElement();
					tag->setFullName(ctx->getNextString()->toString());

					MarkupTagElement* tagCurrent = ctx->getCurrentTag();
					if (tagCurrent != NULL)
					{
						// printf("created new MarkupTagElement named '%s'; added as child of currentElement '%s'\n", nextString->toString(), currentElement->getFullName());
						tagCurrent->addChildElement(tag);
					}
					else
					{
						printf("created new MarkupTagElement named '%s'; added as child of MarkupObjectModel\n", ctx->getNextString()->toString()->toCharArray());
						mom->addChildElement(tag);
					}

					ctx->setCurrentElement(tag);
					ctx->setInsideAttributeArea(true);
				}
				ctx->getNextString()->clear();
				break;
			}
			default:
			{
				ctx->getNextString()->append(ctx->getNextChar());
				break;
			}
		}

		ctx->continueLoop();
	}
}
void XMLDataFormat::saveInternal(ObjectModel* objectModel)
{
	MarkupObjectModel* mom = (MarkupObjectModel*)objectModel;
	if (mom == NULL)
	{
		// TODO: throw a new ObjectModelNotSupported exception here
		printf("ue: XMLDataFormat: saveInternal() - object model not supported\n");
		return;
	}
}

} /* namespace XML */
} /* namespace Markup */
} /* namespace DataFormats */
} /* namespace UniversalEditor */
